promql/generated_parser.y

```diff
@@ -35,120 +36,351 @@ import (
+                        { $$ = $2 }
+                | LEFT_PAREN RIGHT_PAREN
+                        {$$ = Expressions{}}
+                ;
+
+function_call_args: function_call_args COMMA expr
+                        { $$ = append($1.(Expressions), $3.(Expr)) }
+                | expr 
+                        { $$ = Expressions{$1.(Expr)} }
+                ;
+
+/*
+ * Expressions inside parentheses.
+ */
+
+paren_expr      : LEFT_PAREN expr RIGHT_PAREN
+                        { $$ = &ParenExpr{Expr: $2.(Expr)} }
+                ;
+
+/*
+ * Offset modifiers.
+ */
+
+offset_expr: expr OFFSET duration
+                        {
+                        yylex.(*parser).addOffset($1, $3)
+                        $$ = $1
+                        }
+                | expr OFFSET error
+                        { yylex.(*parser).unexpected("offset", "duration") }
+                ;
+
+/*
+ * Subquery and range selectors.
+ */
+
+matrix_selector : expr LEFT_BRACKET duration RIGHT_BRACKET
+                        {
+                        vs, ok := $1.(*VectorSelector)
+                        if !ok{
+                                yylex.(*parser).errorf("matrix selectors only allowed for vector selectors")
+                        }
+                        if vs.Offset != 0{
+                                yylex.(*parser).errorf("no offset modifiers allowed before range selector")
+                        }
+                        $$ = &MatrixSelector{
+                                Name: vs.Name,
+                                Offset: vs.Offset,
+                                LabelMatchers: vs.LabelMatchers,
+                                Range: $3,
+                        }
+                        }
+                ;
+
+subquery_expr   : expr LEFT_BRACKET duration COLON maybe_duration RIGHT_BRACKET
+                        {
+                        $$ = &SubqueryExpr{
+                                Expr:  $1.(Expr),
+                                Range: $3,
+                                Step:  $5,
+                        }
+                        }
+                | expr LEFT_BRACKET duration COLON duration error
+                        { yylex.(*parser).unexpected("subquery selector", "\"]\"") }
+                | expr LEFT_BRACKET duration COLON error
+                        { yylex.(*parser).unexpected("subquery selector", "duration or \"]\"") }
+                | expr LEFT_BRACKET duration error
+                        { yylex.(*parser).unexpected("subquery or matrix selector", "\":\" or \"]\"") }
+                | expr LEFT_BRACKET error
+                        { yylex.(*parser).unexpected("subquery selector", "duration") }
+                ;
 
-label_matchers  : 
-                LEFT_BRACE label_match_list RIGHT_BRACE
+/*
+ * Unary expressions.
+ */
+
+unary_expr      :
+                /* gives the rule the same prec as POW, not a perfect solution */
```

Comment:
Does this matter? I'd expect unary to have the highest precedence.

Reply:
The precedence has to be explicitly specified to make the grammar unambiguous and avoid so called shift/reduce conflicts.

In usual mathematical notation, unary operators actually have the same precedence as multiplication, so `-a^2` evaluates to `-(a^2)`.

However, in PromQL, unary operators have indeed the highest precedence, which was not properly implemented here.

I've rewritten that grammar rule to correct this and added an extra unit test for it.

Reply:
Hmm, so why didn't unittests catch that? We don't document the precedence (or even existence) of the unary operators.

Reply:
There are a bunch of unit tests for operator precedence, which cover most cases.

However the special case of combining an unary operator with `^` slipped trough.

I've tested it with old Prometheus versions and the unary operators do have a higher precedence than `^` there.

Reply:
That sounds like an unintentional mistake, we should follow what other languages typically do here.

Reply:
For this PR I think we should rather not intentionally change PromQL behaviour since it's already complex enough.

I'd prefer doing this as a follow up.

Reply:
I've created an issue since I feel this should be discussed before changing the current behaviour.

https://github.com/prometheus/prometheus/issues/6575

For this PR, I'd like to leave it as is.